File: 399.Evaluate-Division.go
// https://leetcode.com/problems/evaluate-division/description/
//
// algorithms
// Medium (43.63%)
// Total Accepted:    44.9k
// Total Submissions: 102.9k
// beats 100.0% of golang submissions
// 这道题的关键在于把问题抽象化为graph，然后就是简单的dfs了

package leetcode

func calcEquation(equations [][]string, values []float64, queries [][]string) []float64 {
	paramMap, cntIdx := make(map[string]int), 0

	for _, equation := range equations {
		for i := 0; i < 2; i++ {
			if _, ok := paramMap[equation[i]]; !ok {
				paramMap[equation[i]] = cntIdx
				cntIdx++
			}
		}
	}

	graph := make([][]float64, cntIdx)
	for i := 0; i < cntIdx; i++ {
		graph[i] = make([]float64, cntIdx)
	}

	equationLen := len(equations)
	for i := 0; i < equationLen; i++ {
		x, y := paramMap[equations[i][0]], paramMap[equations[i][1]]
		graph[x][y] = values[i]
		graph[y][x] = 1 / values[i]
	}

	queryLen := len(queries)
	res := make([]float64, queryLen)
	for idx, query := range queries {
		start, end := getMapKey(paramMap, query[0]), getMapKey(paramMap, query[1])
		if start == -1 || end == -1 {
			res[idx] = -1.0
		} else {
			isVisited := make([]bool, cntIdx)
			res[idx] = dfsGraph(graph, start, end, 1.0, isVisited)
		}
	}

	return res
}

func getMapKey(m map[string]int, k string) int {
	if v, ok := m[k]; ok {
		return v
	}
	return -1
}

func dfsGraph(graph [][]float64, start int, end int, multiply float64, isVisited []bool) float64 {
	isVisited[start] = true
	if graph[start][end] > 0 {
		return multiply * graph[start][end]
	}
	for idx, n := range graph[start] {
		if n > 0 && !isVisited[idx] {
			res := dfsGraph(graph, idx, end, multiply*n, isVisited)
			if res != -1.0 {
				return res
			}
		}
	}
	return -1.0
}

File: 399.Evaluate-Division.py
# https://leetcode.com/problems/arithmetic-slices/
#
# algorithms
# Medium (48.29%)
# Total Accepted:    88,386
# Total Submissions: 183,034


from collections import defaultdict


class Solution(object):
    def calcEquation(self, equations, values, queries):
        """
        :type equations: List[List[str]]
        :type values: List[float]
        :type queries: List[List[str]]
        :rtype: List[float]
        """
        equations_map = {}
        hash_map = defaultdict(set)

        for idx, [a, b] in enumerate(equations):
            equations_map[(a, b)] = values[idx]
            hash_map[a].add(b)
            hash_map[b].add(a)

        query_len = len(queries)
        res = [-1.0] * query_len

        for i in xrange(query_len):
            a, b = queries[i]
            if a not in hash_map or b not in hash_map:
                continue

            if (a, b) in equations_map:
                res[i] = equations_map[(a, b)]

            elif (b, a) in equations_map:
                res[i] = 1 / equations_map[(b, a)]

            else:
                if a in hash_map:
                    is_visited = {a}
                    r = self.recursive(
                        hash_map, equations_map, a, b, is_visited)
                    if r != float('-inf'):
                        res[i] = r
                if res[i] == -1.0 and b in hash_map:
                    is_visited = {b}
                    r = self.recursive(
                        hash_map, equations_map, b, a, is_visited)
                    if r != float('-inf'):
                        res[i] = 1 / r

                if res[i] != -1.0:
                    equations_map[(a, b)] = res[i]
                    hash_map[a].add(b)
                    hash_map[b].add(a)

        return res

    def recursive(self, hash_map, equations_map, cur, target, is_visited):
        if target in hash_map[cur]:
            return equations_map[(cur, target)] if (cur, target) in equations_map else 1 / equations_map[(target, cur)]

        for t in hash_map[cur]:
            if t not in is_visited:
                v = equations_map[(cur, t)] if (
                    cur, t) in equations_map else 1 / equations_map[(t, cur)]
                return v * self.recursive(hash_map, equations_map, t, target, is_visited | {t})

        return float('-inf')
